File: example-java/runner/src/main/java/Runner.java
import arrow.core.Either;
import io.p8e.ContractManager;
import io.p8e.contracts.examplejava.ExampleContracts.HelloWorldJavaContract;
import io.p8e.exception.P8eError;
import io.p8e.functional.ContractErrorHandler;
import io.p8e.functional.ContractEventHandler;
import io.p8e.proto.ContractSpecs;
import io.p8e.proto.example.HelloWorldExample;
import io.p8e.proxy.Contract;
import io.provenance.p8e.shared.extension.LoggerExtensionsKt;
import java.util.UUID;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import org.slf4j.Logger;

public class Runner {

    public static void main(String[] args) throws InterruptedException {
        Logger log = LoggerExtensionsKt.logger(HelloWorldExample.class.getName());

        String p8eUrl = System.getenv("API_URL");
        String key = System.getenv("PRIVATE_KEY");

        CountDownLatch latch = new CountDownLatch(1);
        AtomicBoolean success = new AtomicBoolean(false);
        UUID executionUuid = UUID.randomUUID();
        UUID scopeUuid = UUID.randomUUID();
        String fistName = "First Name for " + scopeUuid;
        String middleName = "Middle Name for " + scopeUuid;
        String lastName = "Last Name for " + scopeUuid;

        ContractEventHandler<HelloWorldJavaContract> completeHandler = contract -> {
            if(UUID.fromString(contract.getStagedExecutionUuid().getValue()).equals(executionUuid)) {
                log.info("Contract Execution is successful, starting latch countdown.");
                success.set(true);
                latch.countDown();
            } else {
                log.info("ACKing contract from previous run.");
            }
            return true;
        };

        ContractErrorHandler errorHandler = envelopeError -> {
            if(UUID.fromString(envelopeError.getError().getExecutionUuid().getValue()).equals(executionUuid)) {
                log.error("Contract Execution failed, starting latch countdown. " + envelopeError.getError().getMessage());
            }
            return true;
        };

        ContractManager cm = ContractManager.Companion.create(key, p8eUrl, 60_000);
        cm.watchBuilder(HelloWorldJavaContract.class)
                .stepCompletion(completeHandler)
                .error(errorHandler)
                .watch();

        Contract<HelloWorldJavaContract> contract = cm.newContract(
                HelloWorldJavaContract.class,
                scopeUuid,
                executionUuid,
                ContractSpecs.PartyType.OWNER,
                "io.p8e.contracts.examplejava.helloWorld"
        );
        contract.addProposedFact(
                "name",
                HelloWorldExample.ExampleName.newBuilder()
                        .setFirstName(fistName)
                        .setMiddleName(middleName)
                        .setLastName(lastName)
                        .build()
        );

        Either<P8eError, Contract<HelloWorldJavaContract>> result = contract.execute();
        result.map(resultInner -> {
            log.info("Accepted with scope " + scopeUuid + " and execution " + executionUuid);

            boolean latchSuccess = false;

            try {
                latchSuccess = latch.await(120, TimeUnit.SECONDS);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }

            if(latchSuccess && success.get()) {
                log.info("Contract completed successfully!");
            } else if (!latchSuccess) {
                log.error("Contract errored!");
            } else {
                log.error("Contract timed out!");
            }

            return null;
        }).mapLeft(p8eError -> {
            log.error("Error with envelope " + HelloWorldJavaContract.class.getName() + " " + p8eError);
            return null;
        });

        Thread.sleep(2_500);
        cm.close();
    }
}

File: example-kotlin/runner/src/main/kotlin/Runner.kt
import io.p8e.ContractManager
import io.p8e.contracts.examplekotlin.HelloWorldContract
import io.p8e.exception.message
import io.p8e.proto.ContractSpecs
import io.p8e.proto.example.HelloWorldExample
import io.p8e.util.toUuidProv
import io.provenance.p8e.shared.extension.logger
import java.util.UUID
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

fun main() {
    val log = logger()
    val p8eUrl = System.getenv("API_URL")
    val key = System.getenv("PRIVATE_KEY")

    val latch = CountDownLatch(1)
    var success = false
    val executionUuid = UUID.randomUUID()
    val scopeUuid = UUID.randomUUID()
    val firstName = "First Name For $scopeUuid"
    val middleName = "Middle Name For $scopeUuid"
    val lastName = "Last Name For $scopeUuid"

    val cm = ContractManager.create(key, p8eUrl)
    cm.watchBuilder(HelloWorldContract::class.java)
        .stepCompletion {
            if(it.getStagedExecutionUuid().toUuidProv() == executionUuid) {
                log.info("Contract Execution is successful, starting latch countdown.")

                success = true
                latch.countDown()
            } else {
                log.info("ACKing contract from previous run.")
            }

            true
        }
        .error {
            if(it.error.executionUuid.toUuidProv() == executionUuid) {
                log.error("Contract Execution failed, starting latch countdown. ${it.error.message}")

                latch.countDown()
            }
            true
        }
        .also { it.watch() }

    val contract = cm.newContract(
        contractClazz = HelloWorldContract::class.java,
        scopeUuid = scopeUuid,
        executionUuid = executionUuid,
        invokerRole = ContractSpecs.PartyType.OWNER,
    ).apply {
        addProposedFact(
            "name",
            HelloWorldExample.ExampleName.newBuilder()
                .setFirstName(firstName)
                .setMiddleName(middleName)
                .setLastName(lastName)
                .build()
        )
    }

    contract.execute().map {
        log.info("Accepted with scope $scopeUuid and execution $executionUuid")

        val latchSuccess = latch.await(120, TimeUnit.SECONDS)

        if (latchSuccess && success) {
            log.info("Contract completed successfully!")
        } else if (latchSuccess) {
            log.error("Contract errored!")
        } else {
            log.error("Contract timed out!")
        }
    }.mapLeft {
        log.error("Error with envelope ${it::class.java.name} ${it.message()}")
    }

    // allow final ACK to be sent to server
    Thread.sleep(2_500)

    cm.close()
}
